logstash-core/lib/logstash/instrument/metric_store.rb

```diff
@@ -0,0 +1,107 @@
+    #
+    # @param [Array] The path where the values should be located
+    # @param [Object] The default object if the value is not found in the path
+    # @return [Object] Return the new_value of the retrieve object in the tree
+    def fetch_or_store(namespaces, key, default_value = nil)
+      fetch_or_store_namespaces(namespaces).fetch_or_store(key, block_given? ? yield(key) : default_value)
+    end
+    
+    # This method allow to retrieve values for a specific path,
+    #
+    # Also support `*` as a globbing path
+    #
+    # @param [Array] The path where values should be located
+    # @return nil if the values are not found
+    def get(*key_paths)
+      get_recursively(key_paths, @store)
+    end
+
+    # Take all the individuals `MetricType` and convert them to `Logstash::Event`
+    #
+    # @return [Array] An array of all metric transformed in `Logstash::Event`
+    def to_events
+      to_events_recursively(@store).flatten
+    end
+
+    private
+    def get_recursively(key_paths, map)
+      key_candidate = key_paths.shift
+
+      if key_paths.empty?
+        return map[key_candidate]
+      elsif (next_map = map[key_candidate]).is_a?(Concurrent::Map)
+        return get_recursively(key_paths, next_map)
+      else
+        return nil
+      end
+    end
+
+    def to_events_recursively(values)
+      events = []
+      values.each_value do |value|
+        if value.is_a?(Concurrent::Map)
+          events << to_events_recursively(value) 
+        else
+          events << value.to_event
+        end
+      end
+      return events
+    end
+
+    # This method iterate through the namespace path and try to find the corresponding 
+    # value for the path, if the any part of the path is not found it will 
+    # create it.
+    #
+    # @param [Array] The path where values should be located
+    # @raise [ConcurrentMapExpected] Raise if the retrieved object isn't a `Concurrent::Map`
+    # @return [Concurrent::Map] Map where the metrics should be saved
+    def fetch_or_store_namespaces(namespaces_path)
+      path_map = fetch_or_store_namespace_recursively(@store, namespaces_path)
+      
+      # This mean one of the namespace and key are colliding
+      # and we have to deal it upstream.
+      unless path_map.is_a?(Concurrent::Map)
+        raise ConcurrentMapExpectedError, "Expecting a `Concurrent::map`, class:  #{path_map.class.name} for namespaces_path: #{namespaces_path}"
+      end
+
+      return path_map
+    end
+
+    # Recursively fetch or create the namespace paths through the `MetricStove`
+    # This algorithm use an index to known which keys to search in the map.
+    # This doesn't cloning the array if we want to give a better feedback to the user
+    #
+    # @param [Concurrent::Map] Map to search for the key
+    # @param [Array] List of path to create
+    # @param [Fixnum] Which part from the list to create
+    #
+    def fetch_or_store_namespace_recursively(map, namespaces_path, idx = 0)
```

Comment:
not sure I understand the purpose of the ``` idx``` param? I think is also not used, right?